### ISSUE TITLE
UV coordinates incorrect in GLES2 fragment shader when using Nvidia draw_rect flicker workaround

### ISSUE THREAD

**Issue by loganjk** [godotengine/godot#24563]
**Godot version:**
3.1 alpha 4 win64 (Classical)

**OS/device including version:**
Windows 10 Pro, Intel i7-4790K, integrated GPU only (Intel HD Graphics 4600), driver igdkmd64.sys version 20.19.15.5058

**Issue description:**
In an extremely simple 3.1 alpha 4/GLES2 fragment shader, UV coordinates do not behave as expected, instead seemingly being equal to vec2(0.0) everywhere, but this problem does not occur in alpha 4/GLES3 or alpha 2/GLES2.  The shader editor Output window shows the message: `drivers/gles2/shader_gles2.cpp:681 - Condition '!custom_code_map.has(p_code_id) ' is true.`  (I did not test on alpha 3, since the win64 .zip at the download link in the alpha 3 news post seems to actually contain the x11 version: https://downloads.tuxfamily.org/godotengine/3.1/alpha3/Godot_v3.1-alpha3_win64.exe.zip.)

**Steps to reproduce:**
Start a new project.  Create a Node2D with a ColorRect child node with a ShaderMaterial with the following shader:

```
shader_type canvas_item;

void fragment() {
	COLOR=vec4(UV, 1.0, 1.0);
}
```

Instead of the expected color gradient, this is painting the canvas a constant blue (R=G=0.0) everywhere. 
 Under GLES3 or alpha2/GLES 2, however, the shader behaves as expected.

**Minimal reproduction project:**
[Godot Bug Report 2.zip](https://github.com/godotengine/godot/files/2705696/Godot.Bug.Report.2.zip)

**Comment by Zylann** [comment]
UVs used to be zero when no texture is involved in the node being drawn, maybe that happened again?

**Comment by Chaosus** [comment]
That really strange problem(it seems like fixed in nighty/custom Godot builds) but still presents in alpha 4
Need to check in beta

**Comment by loganjk** [comment]
Still a problem in 3.1 beta 1.

**Comment by akien-mga** [comment]
Might be a regression from #23126 (and follow-up #24312).

**Comment by QbieShay** [comment]
Cannot reproduce on latest master ...
![image](https://user-images.githubusercontent.com/7917475/50924099-e5633700-144e-11e9-8163-51810bf0fd09.png)

**Comment by Chaosus** [comment]
Its fixed in latest master indeed, but for some reason it breaks up in release builds(beta, alpha). Still a problem in beta 1.

**Comment by QbieShay** [comment]
The only notable difference that I can see is that UV are calculated from different sources depending if a rect is used

#ifdef USE_TEXTURE_RECT

	if (dst_rect.z < 0.0) { // Transpose is encoded as negative dst_rect.z
		uv_interp = src_rect.xy + abs(src_rect.zw) * vertex.yx;
	} else {
		uv_interp = src_rect.xy + abs(src_rect.zw) * vertex;
	}

	vec4 outvec = vec4(0.0, 0.0, 0.0, 1.0);

	// This is what is done in the GLES 3 bindings and should
	// take care of flipped rects.
	//
	// But it doesn't.
	// I don't know why, will need to investigate further.

	outvec.xy = dst_rect.xy + abs(dst_rect.zw) * select(vertex, vec2(1.0, 1.0) - vertex, lessThan(src_rect.zw, vec2(0.0, 0.0)));

	// outvec.xy = dst_rect.xy + abs(dst_rect.zw) * vertex;
#else
	vec4 outvec = vec4(vertex.xy, 0.0, 1.0);

	uv_interp = uv_attrib;


I won't be able to debug this for a while but you can try putting the uv_interp=uv_attrib out if the ifdef and see if it happens also in debug mode..

**Comment by Chaosus** [comment]
Indeed, its fixed

**Comment by QbieShay** [comment]
How was it fixed? It would be useful to know ..

**Comment by akien-mga** [comment]
Might not be fixed actually but only hidden by #24987. Can you reproduce the issue if you enable `"rendering/quality/2d/gles2_use_nvidia_rect_flicker_workaround"`? I see this code branch does some UV swapping.

**Comment by loganjk** [comment]
Enabling `rendering/quality/2d/gles2_use_nvidia_rect_flicker_workaround` causes the problem to return (after restart), as @akien-mga suspected.
